Add CJK font to support Chinese/Japanese/Korean UML diagram

Miyoshi-Ryota opened the pull request:
To use Chinese/Japanese/Korean in a UML diagram, we need to add a font.

For example, compile the following asciidoctor file including UML diagram in Japanese
```
= This file is a sample file including Japanese in UML

== Sample UML
[plantuml]
----
@startuml

人間A --> 人間B : メッセージの送信
人間B --> 人間A : 返信

@enduml
----
```

Without adding the font, the output will be garbled as shown below.
![without_adding_font](https://user-images.githubusercontent.com/23313273/117149712-6b5ed500-adf2-11eb-8174-5e635b8d3efa.png)

If the font is added, the output will be normal as shown below.
![adding_font](https://user-images.githubusercontent.com/23313273/117149732-7154b600-adf2-11eb-9468-b5aac4be9869.png)

This is a logs to indicate this.
```
mrcsce@pop-os:~/programming/asciidoctor-japanese$ cat sample_including_japanese.adoc 
= This file is a sample file including Japanese in UML

== Sample UML
[plantuml]
----
@startuml

人間A --> 人間B : メッセージの送信
人間B --> 人間A : 返信

@enduml
----
mrcsce@pop-os:~/programming/asciidoctor-japanese$ sudo docker run -it -v /home/mrcsce/programming/asciidoctor-japanese:/documents/ asciidoctor/docker-asciidoctor
bash-5.1# ls
sample_including_japanese.adoc
bash-5.1# asciidoctor -r asciidoctor-diagram -o without_adding_font.html sample_including_japanese.adoc 
bash-5.1# mv diag-de5339e47def289856ddeb2fd653f5d6.png  without_adding_font/
bash-5.1# mv without_adding_font.html without_adding_font
bash-5.1# ls
sample_including_japanese.adoc  without_adding_font
bash-5.1# mkdir adding_font
bash-5.1# apk add font-noto-cjk
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/1) Installing font-noto-cjk (0_git20181130-r1)
Executing fontconfig-2.13.1-r3.trigger
Executing mkfontscale-1.2.1-r1.trigger
OK: 312 MiB in 132 packages
bash-5.1# asciidoctor -r asciidoctor-diagram -o adding_font.html sample_including_japanese.adoc
bash-5.1# ls
adding_font                                diag-de5339e47def289856ddeb2fd653f5d6.png  without_adding_font
adding_font.html                           sample_including_japanese.adoc
bash-5.1# mv adding_font.html adding_font
bash-5.1# mv diag-de5339e47def289856ddeb2fd653f5d6.png adding_font/
bash-5.1# 
```

dduportal commented:
Hello @Miyoshi-Ryota , thank you for this nice contribution! I've qualified this as a bugfix, since it fixes an unwanted behavior: is it ok for you, or do you prefer to qualifies this as an enhancement?

The change looks good, it's ok, however do you mind adding a test in the test harness? You can find an example  of a test for other fonts here: https://github.com/asciidoctor/docker-asciidoctor/blob/main/tests/test_suite.bats#L151 (2 different tests for 2 different fonts).

Miyoshi-Ryota commented:
Thank you for your review!

> I've qualified this as a bugfix, since it fixes an unwanted behavior: is it ok for you, or do you prefer to qualifies this as an enhancement?

That you've qualified this as a bugfix is ok for me.

> The change looks good, it's ok, however do you mind adding a test in the test harness?

Ok! I've added a test case. 

A below is tests result on main branch (without this patch). This is the log that confirms that the test will fail without the patch.
```
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
ubuntu       latest    7e0aa2d69a15   12 days ago   72.7MB
alpine       latest    6dbb9cc54074   3 weeks ago   5.61MB
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make clean
rm -rf "/home/mrcsce/programming/docker-asciidoctor/cache"
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make build
<...Omit build logs...>
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo docker images
REPOSITORY                       TAG                   IMAGE ID       CREATED          SIZE
asciidoctor/docker-asciidoctor   1.6.0                 7ef7838c66da   48 minutes ago   455MB
asciidoctor/docker-asciidoctor   1.6.0-build-haskell   dc43faa92f0e   51 minutes ago   3.07GB
asciidoctor/docker-asciidoctor   1.6.0-minimal         955049e3fcd6   56 minutes ago   38MB
ubuntu                           latest                7e0aa2d69a15   12 days ago      72.7MB
alpine                           latest                6dbb9cc54074   3 weeks ago      5.61MB
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make test 
bats /home/mrcsce/programming/docker-asciidoctor/tests/*.bats
 ✓ The Docker image to test is available
 ✓ asciidoctor is installed and in version 2.0.15
 ✓ Timezone data is present in the image
 ✓ asciidoctor-pdf is installed and in version 1.5.4
 ✓ asciidoctor-revealjs is callable without error
 ✓ make is installed and in the path
 ✓ asciidoctor-epub3 is installed and in the path
 ✓ asciidoctor-fb2 is installed and in version 0.5.1
 ✓ curl is installed and in the path
 ✓ bash is installed and in the path
 ✓ java is installed, in the path, and executable
 ✓ python3 is installed, in the path, and executable
 ✓ dot (from Graphviz) is installed and in the path
 ✓ asciidoctor-confluence is installed and in the path
 ✓ kramdoc is installed and in the path
 ✓ git command line tool is installed and in the path
 ✓ We can generate an HTML document from basic example
 ✓ We can generate a PDF document from basic example
 ✓ We can generate an FB2 document from basic example without errors/warnings
 ✓ We can generate an HTML document with a diagram with asciidoctor-diagram as backend
 ✓ Bakoma Fonts are installed to render correctly the square root from asciidoctor-mathematical
 ✗ Noto CJK Fonts are installed to render correctly the PlantUML diagram from asciidoctor-diagram
   (in test file tests/test_suite.bats, line 156)
     `docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" apk info font-noto-cjk' failed
   WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.13/main: No such file or directory
   WARNING: Ignoring https://dl-cdn.alpinelinux.org/alpine/v3.13/community: No such file or directory
 ✓ DejaVu Fonts are installed to get corretly rendered PlantUML-Graphs
 ✓ We can generate an HTML document with asciidoctor-mathematical as backend
 ✓ We can generate an EPub document with asciidoctor-epub3
 ✓ We can generate an HTML document with asciimath as backend
 ✓ We can generate a PDF document with asciidoctor-mathematical as backend
 ✓ We can generate a Reveal.js Slide deck
 ✓ We can generate HTML documents with different syntax-colored codes
 ✓ We can generate PDF documents with different syntax-colored codes
 ✓ We can convert a Markdown file to an AsciiDoc file
 ✓ We can produce a website with citations from bibtex

32 tests, 1 failure

make: *** [Makefile:51: test] エラー 1
mrcsce@pop-os:~/programming/docker-asciidoctor$ 
mrcsce@pop-os:~/programming/docker-asciidoctor$ git diff
diff --git a/tests/test_suite.bats b/tests/test_suite.bats
index d295c12..81e4364 100644
--- a/tests/test_suite.bats
+++ b/tests/test_suite.bats
@@ -152,6 +152,10 @@ teardown() {
   docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" apk info font-bakoma-ttf
 }
 
+@test "Noto CJK Fonts are installed to render correctly the PlantUML diagram from asciidoctor-diagram" {
+  docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" apk info font-noto-cjk
+}
+
 @test "DejaVu Fonts are installed to get corretly rendered PlantUML-Graphs" {
   docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" fc-list "DejaVu Sans"
 }
mrcsce@pop-os:~/programming/docker-asciidoctor$ 
mrcsce@pop-os:~/programming/docker-asciidoctor$ git branch
  Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container
* main
mrcsce@pop-os:~/programming/docker-asciidoctor$ 
```

And then below is tests result on the branch (with this patch).
```
mrcsce@pop-os:~/programming/docker-asciidoctor$ git stash
Saved working directory and index state WIP on main: 4fc4833 Merge pull request #200 from asciidoctor/updatecli_21599d24a9c8717a80dbae70d0988924634fa619f9e704f7083e27e36a700897
mrcsce@pop-os:~/programming/docker-asciidoctor$ git checkout Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container 
Switched to branch 'Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container'
Your branch is up to date with 'origin/Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container'.
mrcsce@pop-os:~/programming/docker-asciidoctor$ git stash pop
ブランチ Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container
Your branch is up to date with 'origin/Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   tests/test_suite.bats

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (a0cea554549a2f841d589317d96f82caf82fdbed)
mrcsce@pop-os:~/programming/docker-asciidoctor$ git diff
diff --git a/tests/test_suite.bats b/tests/test_suite.bats
index d295c12..81e4364 100644
--- a/tests/test_suite.bats
+++ b/tests/test_suite.bats
@@ -152,6 +152,10 @@ teardown() {
   docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" apk info font-bakoma-ttf
 }
 
+@test "Noto CJK Fonts are installed to render correctly the PlantUML diagram from asciidoctor-diagram" {
+  docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" apk info font-noto-cjk
+}
+
 @test "DejaVu Fonts are installed to get corretly rendered PlantUML-Graphs" {
   docker run -t --rm "${DOCKER_IMAGE_NAME_TO_TEST}" fc-list "DejaVu Sans"
 }
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make clean
rm -rf "/home/mrcsce/programming/docker-asciidoctor/cache"
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make build
<...Omit build logs...>
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo docker images
REPOSITORY                       TAG                                                                            IMAGE ID       CREATED          SIZE
asciidoctor/docker-asciidoctor   Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container                 32370c4e94d5   19 minutes ago   547MB
asciidoctor/docker-asciidoctor   1.6.0                                                                          7ef7838c66da   51 minutes ago   455MB
asciidoctor/docker-asciidoctor   1.6.0-build-haskell                                                            dc43faa92f0e   54 minutes ago   3.07GB
asciidoctor/docker-asciidoctor   Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container-build-haskell   dc43faa92f0e   54 minutes ago   3.07GB
asciidoctor/docker-asciidoctor   1.6.0-minimal                                                                  955049e3fcd6   59 minutes ago   38MB
asciidoctor/docker-asciidoctor   Add-CJK-font-to-support-Japanese-UML-diagram-on-this-container-minimal         955049e3fcd6   59 minutes ago   38MB
ubuntu                           latest                                                                         7e0aa2d69a15   12 days ago      72.7MB
alpine                           latest                                                                         6dbb9cc54074   3 weeks ago      5.61MB
mrcsce@pop-os:~/programming/docker-asciidoctor$ 
mrcsce@pop-os:~/programming/docker-asciidoctor$ sudo make test 
bats /home/mrcsce/programming/docker-asciidoctor/tests/*.bats
 ✓ The Docker image to test is available
 ✓ asciidoctor is installed and in version 2.0.15
 ✓ Timezone data is present in the image
 ✓ asciidoctor-pdf is installed and in version 1.5.4
 ✓ asciidoctor-revealjs is callable without error
 ✓ make is installed and in the path
 ✓ asciidoctor-epub3 is installed and in the path
 ✓ asciidoctor-fb2 is installed and in version 0.5.1
 ✓ curl is installed and in the path
 ✓ bash is installed and in the path
 ✓ java is installed, in the path, and executable
 ✓ python3 is installed, in the path, and executable
 ✓ dot (from Graphviz) is installed and in the path
 ✓ asciidoctor-confluence is installed and in the path
 ✓ kramdoc is installed and in the path
 ✓ git command line tool is installed and in the path
 ✓ We can generate an HTML document from basic example
 ✓ We can generate a PDF document from basic example
 ✓ We can generate an FB2 document from basic example without errors/warnings
 ✓ We can generate an HTML document with a diagram with asciidoctor-diagram as backend
 ✓ Bakoma Fonts are installed to render correctly the square root from asciidoctor-mathematical
 ✓ Noto CJK Fonts are installed to render correctly the PlantUML diagram from asciidoctor-diagram
 ✓ DejaVu Fonts are installed to get corretly rendered PlantUML-Graphs
 ✓ We can generate an HTML document with asciidoctor-mathematical as backend
 ✓ We can generate an EPub document with asciidoctor-epub3
 ✓ We can generate an HTML document with asciimath as backend
 ✓ We can generate a PDF document with asciidoctor-mathematical as backend
 ✓ We can generate a Reveal.js Slide deck
 ✓ We can generate HTML documents with different syntax-colored codes
 ✓ We can generate PDF documents with different syntax-colored codes
 ✓ We can convert a Markdown file to an AsciiDoc file
 ✓ We can produce a website with citations from bibtex

32 tests, 0 failures

mrcsce@pop-os:~/programming/docker-asciidoctor$ 
```

dduportal commented:
That is a nice job! Thanks a lot!
